src/v/cluster/id_allocator_stm.cc

```diff
@@ -32,232 +33,171 @@ static model::record_batch serialize_cmd(T t, model::record_batch_type type) {
     auto r = b.copy_records();
     auto& record = *r.begin();
     auto rk = reflection::adl<uint8_t>{}.from(record.release_key());
 
+    _insync_offset = b.last_offset();
+
     if (rk == allocation_cmd::record_key) {
         allocation_cmd cmd = reflection::adl<allocation_cmd>{}.from(
           record.release_value());
         if (_should_cache) {
-            _cache.emplace_back(cached_allocation_cmd{b.last_offset(), cmd});
+            _cache.emplace_back(cmd);
         } else {
-            execute(b.last_offset(), cmd);
+            _state += cmd.range;
         }
     } else if (rk == prepare_truncation_cmd::record_key) {
-        prepare_truncation_cmd cmd = reflection::adl<prepare_truncation_cmd>{}
-                                       .from(record.release_value());
         if (!_should_cache) {
-            _processed = 0;
             _should_cache = true;
-            _prepare_state = _state;
             _prepare_offset = b.last_offset();
         }
-        if (auto it = _prepare_promises.find(cmd.seq);
-            it != _prepare_promises.end()) {
-            it->second.set_value(true);
-        }
     } else if (rk == execute_truncation_cmd::record_key) {
         execute_truncation_cmd cmd = reflection::adl<execute_truncation_cmd>{}
                                        .from(record.release_value());
         if (_should_cache && _prepare_offset == cmd.prepare_offset) {
             _state = cmd.state;
-            for (auto& it : _cache) {
-                execute(it.offset, it.cmd);
+            for (auto& cmd : _cache) {
+                _state += cmd.range;
             }
             _cache.clear();
             _should_cache = false;
-            return _c->write_last_applied(_prepare_offset).then([this] {
-                return _c->write_snapshot(raft::write_snapshot_cfg(
-                  model::offset(_prepare_offset - 1),
-                  iobuf(),
-                  raft::write_snapshot_cfg::should_prefix_truncate::yes));
-            });
         }
-    }
-    return ss::now();
-}
-
-void id_allocator_stm::execute(model::offset offset, allocation_cmd c) {
-    auto base = _state;
-    _state += c.range;
-    _processed += 1;
+    } else if (rk == state_cmd::record_key) {
+        _procesing_legacy = false;
+        state_cmd cmd = reflection::adl<state_cmd>{}.from(
+          record.release_value());
+        _state = cmd.next_state;
 
-    id_allocator_stm::log_allocation_result result{
-      c.seq, base, c.range, raft::errc::success, offset};
+        if (_next_snapshot() < 0) {
+            _next_snapshot = _insync_offset;
+            _processed = 0;
+        }
 
-    if (auto it = _promises.find(result.seq); it != _promises.end()) {
-        it->second.set_value(result);
+        if (_c->is_leader()) {
+            _processed++;
+            if (_processed > _log_capacity) {
+                return _c
+                  ->write_snapshot(raft::write_snapshot_cfg(
```

Comment:
Does this mean that on followers log can grow unbounded?

Reply:
Good point! I wrongly assumed that we replicate log truncation events.